theano/tensor/opt.py

```diff
@@ -5696,11 +5696,24 @@ def local_opt_alloc(node):
                                                     only_process_constants=True)
                     assert val.size == 1
                     # check which type of op
-                    casted = T.mul(*shapes).astype(str(input.dtype))
+                    size = T.mul(*shapes)
+                    if input.dtype == "float32":
+                        # shapes are ints and normally int64.
+                        # We don't want to have a float64 upcast here
+                        # if input is a float32.
+                        size = size.astype(input.dtype)
```

Comment:
Why to this cast here when we reduced on all axes, but not below when we reduce on only some of them?